mule-classloader-model/src/main/java/org/mule/tools/api/classloader/model/ApplicationClassloaderModel.java

```diff
@@ -28,46 +30,40 @@ public ClassLoaderModel getClassLoaderModel() {
     return classLoaderModel;
   }
 
-  public void addMulePluginClassloaderModel(ClassLoaderModel mulePluginClassloaderModel) {
-    this.mulePluginsClassloaderModels.add(mulePluginClassloaderModel);
+  protected Map<ArtifactCoordinates, ClassLoaderModel> getNestedClassLoaderModels() {
+    return nestedClassLoaderModels;
   }
 
-  public void addAllMulePluginClassloaderModels(Collection<ClassLoaderModel> mulePluginClassloaderModels) {
-    this.mulePluginsClassloaderModels.addAll(mulePluginClassloaderModels);
-  }
+  public void mergeDependencies(Collection<ClassLoaderModel> otherClassloaderModels) {
+    for (ClassLoaderModel otherClassLoaderModel : otherClassloaderModels) {
+      Artifact pluginArtifact = getClassLoaderModel().getDependencies().stream()
+          .filter(dependency -> dependency.getArtifactCoordinates().equals(otherClassLoaderModel.getArtifactCoordinates()))
+          .findFirst()
+          .orElseThrow(() -> new IllegalStateException(format("Couldn't find an artifact coordinate for '%s' to merge dependencies as inline",
+                                                              otherClassLoaderModel.getArtifactCoordinates())));
 
-  public Set<Artifact> getArtifacts() {
-    Set<Artifact> artifacts = new HashSet<>();
-    artifacts.addAll(classLoaderModel.getArtifacts());
-    artifacts.addAll(mulePluginsClassloaderModels.stream()
-        .map(ClassLoaderModel::getArtifacts)
-        .flatMap(Collection::stream)
-        .collect(Collectors.toList()));
-    artifacts.addAll(ramlsClassloaderModels.stream()
-        .map(ClassLoaderModel::getArtifacts)
-        .flatMap(Collection::stream)
-        .collect(Collectors.toList()));
-    return artifacts;
-  }
-
-  public List<ClassLoaderModel> getMulePluginsClassloaderModels() {
-    return mulePluginsClassloaderModels;
+      if (nestedClassLoaderModels.putIfAbsent(otherClassLoaderModel.getArtifactCoordinates(), otherClassLoaderModel) != null) {
+        throw new IllegalArgumentException(format("Duplicated definition of a nested class loader model for artifact coordinates '%s'",
```

Comment:
If maven resolution is correct, this should never happen right?

Reply:
Correct, but in case if Maven Client has a bug we are throwing this error here.